src/tree.rs

```diff
@@ -1,15 +1,103 @@
+use crate::{
+    constants::{MERKLE_CRH_PERSONALIZATION, MERKLE_DEPTH_ORCHARD},
+    note::commitment::ExtractedNoteCommitment,
+    primitives::sinsemilla::{i2lebsp_k, HashDomain, K},
+};
+use pasta_curves::{arithmetic::FieldExt, pallas};
+
+use ff::{PrimeField, PrimeFieldBits};
 use rand::RngCore;
+use std::{convert::TryInto, iter};
 
 /// The root of an Orchard commitment tree.
-#[derive(Clone, Debug)]
+#[derive(Eq, PartialEq, Clone, Debug)]
 pub struct Anchor(pub [u8; 32]);
```

Comment:
Normally we implement type constraints as non-contextual checks. I would do so here, even if it would also be rejected later by the contextual check against a previous anchor.